File: engine/b/src/main/java/net/crazyinvestor/engine_b/handler/BithumbWSResponseHandler.java
package net.crazyinvestor.engine_b.handler;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import net.crazyinvestor.engine_b.dto.response.BithumbWSOrderbookDepthResponse;
import net.crazyinvestor.engine_b.dto.response.BithumbWSTickerResponse;
import net.crazyinvestor.engine_b.dto.response.BithumbWSTransactionResponse;
import net.crazyinvestor.engine_b.enums.BithumbOpType;
import net.crazyinvestor.engine_b.event.BithumbOrderbookDepthEvent;
import net.crazyinvestor.engine_b.event.BithumbTickerEvent;
import net.crazyinvestor.engine_b.event.BithumbTransactionEvent;
import net.crazyinvestor.engine_b.service.TickerService;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Component;

import java.util.Objects;

@Component
public class BithumbWSResponseHandler {
    private final TickerService tickerService;
    private final ObjectMapper objectMapper;
    private final ApplicationEventPublisher eventPublisher;

    public BithumbWSResponseHandler(
            final TickerService tickerService,
            final ApplicationEventPublisher eventPublisher
    ) {
        this.tickerService = tickerService;
        this.eventPublisher = eventPublisher;

        this.objectMapper = new ObjectMapper();
        this.objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    public void handle(String data) {
        try {
            JsonNode response = objectMapper.readTree(data);

            if (Objects.nonNull(response.get("status"))) {
                handleStatusMessage(response);
            } else {
                String type = response.get("type").asText();

                if (BithumbOpType.TICKER.getOpType().equals(type)) {
                    handleTickerData(response);
                } else if (BithumbOpType.TRANSACTION.getOpType().equals(type)) {
                    handleTransactionData(response);
                } else if (BithumbOpType.ORDERBOOK_DEPTH.getOpType().equals(type)) {
                    handleOrderbookDepthData(response);
                }
            }
        } catch (JsonProcessingException e) {
            e.printStackTrace();
            System.exit(1);
        }
    }

    private void handleOrderbookDepthData(JsonNode response) throws JsonProcessingException {
        String type = response.get("type").asText();
        JsonNode content = response.get("content");
        Long datetime = content.get("datetime").asLong();

        BithumbWSOrderbookDepthResponse responseObj = objectMapper.treeToValue(content, BithumbWSOrderbookDepthResponse.class);
        responseObj.setDatetime(datetime);
        eventPublisher.publishEvent(new BithumbOrderbookDepthEvent(responseObj));

        // TODO: OrderbookDepth MySQL 저장 로직 필요
    }

    private void handleTransactionData(JsonNode response) throws JsonProcessingException {
        String type = response.get("type").asText();
        JsonNode content = response.get("content");

        BithumbWSTransactionResponse responseObj = objectMapper.treeToValue(content, BithumbWSTransactionResponse.class);
        eventPublisher.publishEvent(new BithumbTransactionEvent(responseObj));

        // TODO: Transaction MySQL 저장 로직 필요
    }

    private void handleTickerData(JsonNode response) throws JsonProcessingException {
        String type = response.get("type").asText();
        JsonNode content = response.get("content");

        BithumbWSTickerResponse responseObj = objectMapper.treeToValue(content, BithumbWSTickerResponse.class);
        eventPublisher.publishEvent(new BithumbTickerEvent(responseObj));

        tickerService.saveBithumbTicker(responseObj);
    }

    private void handleStatusMessage(JsonNode response) {
        String status = response.get("status").asText();
        String resmsg = response.get("resmsg").asText();

        System.out.println("Status response = " + response);
    }
}

File: engine/a/ticker_producer_engine/src/main/kotlin/net/crazyinvestor/ticker_producer_engine/hadler/BithumbWSResponseHandler.kt
package net.crazyinvestor.ticker_producer_engine.hadler

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import net.crazyinvestor.ticker_producer_engine.dto.bithumb.response.BithumbWSOrderbookDepthResponse
import net.crazyinvestor.ticker_producer_engine.dto.bithumb.response.BithumbWSTickerResponse
import net.crazyinvestor.ticker_producer_engine.dto.bithumb.response.BithumbWSTransactionResponse
import net.crazyinvestor.ticker_producer_engine.dto.bithumb.response.BithumbWSTransactionResponseContent
import net.crazyinvestor.ticker_producer_engine.enums.BithumbOpType
import net.crazyinvestor.ticker_producer_engine.service.TickerOpService
import net.crazyinvestor.ticker_producer_engine.service.TickerTxOpService
import org.springframework.context.ApplicationEventPublisher
import org.springframework.stereotype.Component
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toFlux
import reactor.kotlin.core.publisher.toMono
import java.util.*

@Component
class BithumbWSResponseHandler(
    private val tickerService: TickerOpService,
    private val tickerTxOpService: TickerTxOpService,
    private val eventPublisher: ApplicationEventPublisher
) {
    private val objectMapper: ObjectMapper = ObjectMapper()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    private fun readTreeMono(data: String?): Mono<JsonNode> {
        return data.toMono()
            .map { objectMapper.readTree(it) }
            .doOnError { it.printStackTrace(); System.exit(1) }
    }

    fun handle(data: String?) {
        readTreeMono(data)
            .doOnNext { response ->
                when {
                    Objects.nonNull(response["status"]) -> handleStatusMessage(response)
                    else -> when(response["type"].asText()) {
                        BithumbOpType.TICKER.opType -> handleTickerData(response)
                        BithumbOpType.TRANSACTION.opType -> handleTransactionData(response)
                        BithumbOpType.ORDERBOOK_DEPTH.opType -> handleOrderbookDepthData(response)
                        else -> { println("NOT EXISTS TYPE"); }
                    }
                }
            }
            .subscribe()
    }
//        try {
//            val response = objectMapper.readTree(data)
//            when {
//                Objects.nonNull(response["status"]) -> handleStatusMessage(response)
//                else -> when(response["type"].asText()) {
//                    BithumbOpType.TICKER.opType -> handleTickerData(response)
//                    BithumbOpType.TRANSACTION.opType -> handleTransactionData(response)
//                    BithumbOpType.ORDERBOOK_DEPTH.opType -> handleOrderbookDepthData(response)
//                    else -> { println("NOT EXISTS TYPE"); }
//                }
//            }
//        } catch (e: JsonProcessingException) {
//            e.printStackTrace()
//            System.exit(1)
//        }

    @Throws(JsonProcessingException::class)
    private fun handleOrderbookDepthData(response: JsonNode) {
        val type = response["type"].asText()
        val content = response["content"]
        val datetime = content["datetime"].asLong()
        val responseObj: BithumbWSOrderbookDepthResponse = objectMapper.treeToValue<BithumbWSOrderbookDepthResponse>(
            content,
            BithumbWSOrderbookDepthResponse::class.java
        )
        responseObj.datetime = datetime
        // eventPublisher.publishEvent(BithumbOrderbookDepthEvent(responseObj))

        // TODO: OrderbookDepth 저장 로직
    }

    @Throws(JsonProcessingException::class)
    private fun handleTransactionData(response: JsonNode) {
        val type = response["type"].asText()
        val content = response["content"]
        val responseObj: BithumbWSTransactionResponse = objectMapper.treeToValue<BithumbWSTransactionResponse>(
            content,
            BithumbWSTransactionResponse::class.java
        )
        responseObj.list!!
            .toFlux()
            .flatMap { Mono.just(it) }
            .doOnNext {
                tickerTxOpService.produce(it)
            }
            .subscribe()
    }

    private fun handleTickerData(response: JsonNode) {
        // val type = response["type"].asText()
        val content = response["content"]
        val responseObj: BithumbWSTickerResponse = objectMapper.treeToValue<BithumbWSTickerResponse>(
            content,
            BithumbWSTickerResponse::class.java
        )
        tickerService.saveBithumbTicker(responseObj)
    }

    private fun handleStatusMessage(response: JsonNode) {
        val status = response["status"].asText()
        val resmsg = response["resmsg"].asText()
        println("Status response = $response")
    }
}
